File: tradingcards-plugin/src/main/java/net/tinetwork/tradingcards/tradingcardsplugin/managers/Manager.java
package net.tinetwork.tradingcards.tradingcardsplugin.managers;

import com.github.benmanes.caffeine.cache.LoadingCache;

import net.tinetwork.tradingcards.api.manager.Cacheable;
import net.tinetwork.tradingcards.tradingcardsplugin.TradingCards;

import java.util.List;

/**
 * @author sarhatabaot
 */
public abstract class Manager<R,T> implements Cacheable<R,T> {
    protected final TradingCards plugin;
    protected final LoadingCache<R,T> cache;

    protected Manager(final TradingCards plugin) {
        this.plugin = plugin;
        this.cache = loadCache();
        preLoadCache();
    }

    public void preLoadCache() {
        this.cache.getAll(getKeys());
    }

    @Override
    public void forceCacheRefresh() {
        this.cache.invalidateAll();
        preLoadCache();
    }

    public abstract List<R> getKeys();

    public LoadingCache<R, T> getCache() {
        return cache;
    }


}

File: tradingcards-plugin/src/main/kotlin/com/github/sarhatabaot/tradingcards/managers/Manager.kt
package com.github.sarhatabaot.tradingcards.managers

import com.github.benmanes.caffeine.cache.LoadingCache
import net.tinetwork.tradingcards.api.manager.Cacheable
import net.tinetwork.tradingcards.tradingcardsplugin.TradingCards

/**
 *
 * @author sarhatabaot
 */

abstract class Manager<R, T> protected constructor(protected var plugin: TradingCards) :
    Cacheable<R, T> {
    protected val cache: LoadingCache<R, T> by lazy { loadCache() }

    init {
        preLoadCache()
    }

    override fun preLoadCache() {
        cache.getAll(getKeys())
    }

    override fun forceCacheRefresh() {
        cache.invalidateAll()
        preLoadCache()
    }

    abstract fun getKeys(): List<R>?
    fun getCache(): LoadingCache<R, T> {
        return cache
    }
}
